narayana/ArjunaJTA/jta/tests/classes/io/narayana/perf/product/ProductComparison.java

```diff
@@ -107,6 +109,19 @@ public void test2() throws Exception {
         System.out.printf("Completed %d out of %d workloads%n", completionCount.get(), TASK_COUNT);
     }
 
+    private void startWork() throws Exception {
+        try{
+            doWork(worker.getProduct());
+        } catch (Exception e){
```

Comment:
It is defined that JMH requires a BenchmarkException to be thrown?

Reply:
Assuming it is, I wonder why something here https://github.com/jbosstm/performance/blob/master/narayana/ArjunaJTA/jta/tests/classes/io/narayana/perf/product/ProductComparison.java#L134 would not be triggering? Perhaps that can be improved in there? Or if not maybe this clause could be evaluated for moving to there. I would note that the existing protection seems to use an AtomicInteger, rather than a primitive data type.

Looking at doWork a bit more, perhaps this additional code would delay reporting the BenchmarkException that is thrown there?

Reply:
If we want to interrupt the Benchmark during execution, throwing BenchmarkException would do so during runtime. Can't say why in CI, this https://github.com/jbosstm/performance/blob/master/narayana/ArjunaJTA/jta/tests/classes/io/narayana/perf/product/ProductComparison.java#L131 condition is not executing properly. I guess somehow the value of AtomicInteger *errorCount* is not incrementing. I am trying to investigate.

Reply:
We should consider remove tolerance for Exceptions (the bit where errorCount is checked [although an Exception is not an Error in terms of Java package hierarchy]) from our benchmark and just throw the Exception immediately. Perhaps there is a way to configure JMH to deal with error tolerance itself but if not then a failed run can be re-ran.

AtomicInteger seems like it should be thread-safe by being within the java.util.concurrent.atomic package if you look at: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/package-summary.html so I don't really see why it would not get incremented if an exception was thrown.

Reply:
yes it is working correctly. I got confused with something else. Sorry